### PR TITLE
fix: align AKS memory reservations across provisioning modes

### LINKED ISSUE #1761: Node Allocatable resources computation
### Tell us about your request

Hi team 👋 

I was looking into how node allocatable resources—specifically kubeReserved memory—are computed, and I noticed a discrepancy between this provider and standard non-Karpenter AKS nodes.

According to the [AKS node resource reservations documentation](https://learn.microsoft.com/en-us/azure/aks/node-resource-reservations#memory-reservations), `kubeReserved.memory` is calculated dynamically based on `maxPods`:

> 20 MB * Max Pods supported on the Node + 50 MB (or 25% of the total system memory resources).

For context, [karpenter-provider-aws](https://github.com/aws/karpenter-provider-aws) also uses `maxPods` to compute this: [reference](https://github.com/aws/karpenter-provider-aws/blob/1de6d8c602f910a87976d61f1d3a1fb5d7eea726/pkg/providers/instancetype/types.go#L524).

However, in karpenter-provider-azure, it looks like we are using static [tax brackets](https://github.com/Azure/karpenter-provider-azure/blob/66c30d88eb719287f83611a27a9c7541aede2b25/pkg/providers/instancetype/instancetype.go#L47-L88). This means `kubeReserved.memory` is computed purely based on memory capacity, without taking `maxPods` into account.

I'm opening this issue to better understand the context behind this design choice. Are there specific technical limitations driving this?

If not, would the maintainers be open to aligning this logic with the official AKS documentation? Relying on the same calculation would make things much easier and more predictable for users transitioning to Karpenter on Azure.

Thanks for your time and for the work on this provider!

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

The main problem is the inconsistency in capacity planning when running Karpenter alongside (or migrating from) standard AKS node pools. Because karpenter-provider-azure uses a different mathematical model for kubeReserved, the exact same VM SKU yields a different amount of allocatable memory.

Goal: Align the allocatable resource calculations with standard AKS to ensure a seamless, predictable, and fully documented experience for users adopting Karpenter on Azure.

### Are you currently working around this issue?

No

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by GuetaHen**
Fixes #1761

**Description**

### Why we are doing this

When moving from fixed-size AKS node pools to Karpenter/NAP, the provisioner must predict which VM sizes can fit each workload batch. The old non-hardened forecast still uses pre-1.29 memory tax brackets and a 750 MiB eviction threshold, even when AKS bootstraps the node with the newer maxPods-based reservation and a 100 MiB threshold.

That mismatch can exclude a genuinely usable, smaller instance type before cost optimization can select it. Karpenter can then provision a larger node or a combination of nodes, despite unchanged pod requests. The same incorrect forecast can reject a useful replacement during consolidation. This is an input-accuracy problem, not evidence that every small-node allocation or the consolidation algorithm is wrong.

### What changes

Align all non-hardened provisioning modes with the [documented AKS memory reservations](https://learn.microsoft.com/azure/aks/node-resource-reservations#memory-reservations):

- Kube-reserved memory: `min(20 * maxPods + 50, totalMemoryMiB / 4)` MiB.
- Hard memory eviction threshold: `100Mi`, rather than `750Mi`.

Following @theunrepentantgeek and @matthchr's feedback, the calculation now lives directly in `KubeReservedResources`, and `EvictionThreshold` uses the shared 100 MiB default. The old memory brackets, Kubernetes-version selector, policy cache flag, and post-construction overrides are removed. There are only two policies: standard and hardened. `maxPods` already participates in the instance-type cache key.

This applies to `aksscriptless`, `bootstrappingclient`, `aksmachineapi`, and `aksmachineapiheaderbatch`, without a Kubernetes-version gate. **Scriptless now uses the same values both in its forecast and in the generated kubelet configuration for newly created nodes**, including the base eviction fallback. The other modes retain their AKS-owned bootstrap defaults; this aligns the provider's prediction with those defaults. Existing nodes are not reconfigured.

Node-hardening policy, CPU reservations, system reservations, ephemeral-storage eviction accounting, and the separate VM-memory safety margin are unchanged. Pricing and consolidation policy are also unchanged.

### Concrete packing case (synthetic)

The tests use identical **10-CPU/42-GiB workload pods**, plus **1 CPU/3 GiB of per-node overhead**, maxPods=250, and the unchanged 7.5% VM-memory safety margin. These are synthetic inputs, not production workload or billing measurements.

| Instance type | Old predicted allocatable memory | Corrected prediction | Maximum workload pods, before / after |
|---|---:|---:|---:|
| D16as_v6 (64 GiB) | 52.99 GiB | 54.17 GiB | 1 / 1 |
| D32as_v6 (128 GiB) | 108.35 GiB | 113.37 GiB | 2 / 2 |
| D48as_v6 (192 GiB) | 166.27 GiB | 172.57 GiB | 3 / 4 |
| D64as_v6 (256 GiB) | 224.19 GiB | 231.77 GiB | 5 / 5 |
| D96as_v6 (384 GiB) | 340.03 GiB | 350.17 GiB | 8 / 8 |

Four workload pods on one node require **41 CPU and 171 GiB** after the per-node overhead. The old forecast rejects D48 because 171 GiB exceeds 166.27 GiB; D64 is memory-feasible. The corrected D48 forecast accepts the same batch without changing pod requests or lowering the VM-memory safety margin.

If a compatible, available D48 offering is cheaper than D64, this makes a **48-vCPU node eligible instead of a 64-vCPU node: 25% fewer provisioned vCPUs for this example**. Four D16s or two D32s also total 64 vCPUs for these four pods; the correction makes a single D48 a resource-feasible replacement. Raising a minimum SKU alone would not repair the rejected D48 prediction.

### Expected improvement and limits

The expected improvement is more accurate scale-up and replacement choices: fewer artificially rejected instance types and an opportunity to use less provisioned capacity for the same requests. Initialized nodes already use Kubernetes-reported allocatable resources. This does not unlock additional physical memory or reconfigure running nodes; newly created scriptless nodes receive the updated reservations.

The 25% figure is a **conditional resource-fit example, not a guaranteed fleet-wide or billing saving**. Actual choices still depend on batch composition, compatible offerings and prices, available Azure capacity, NodeOverlays, affinities/topology, PDBs, and disruption budgets. This does not promise to eliminate D16 nodes, fix application memory leaks, or resolve every consolidation timeout.

After integration, a live AKS canary should compare predicted versus reported allocatable memory and provisioned vCPUs for comparable workload requests, while monitoring pending pods, evictions/OOMs, and node churn. This PR has not been deployed to managed NAP.

**How was this change tested?**

Ran the complete affected packages as Linux/amd64 test binaries in WSL, with local Kubernetes envtest 1.34.1 and fake Azure APIs:

| Package | Standard Go test functions passed | Ginkgo cases passed |
|---|---:|---:|
| `pkg/providers/instancetype` | 11 | 249 |
| `pkg/providers/imagefamily` | 19 | 41 |
| `pkg/providers/instance` | 7 | 151 |
| `pkg/cloudprovider` | 3 | 155 |
| `pkg/providers/imagefamily/bootstrap` | 8 | - |

**596 Ginkgo cases, 48 standard Go test functions, and 125 Go subtests passed.** There are three existing skipped Ginkgo cases and one existing pending case; the four Ginkgo suite entrypoints are not double-counted as unit tests. After the final test-only follow-up removed the obsolete four-case version/cache table, the complete instance-type suite was rerun; the other four packages are unchanged from the earlier full validation.

* Red/green regression: before changing production code, 11 reservation subtests and the base scriptless eviction test failed against the legacy shared defaults; they pass with the correction.
* Direct shared-helper tests cover maxPods 30/110/250, the 25% cap and rounding boundary. Constructor tests cover all four provisioning modes, hardening precedence, unchanged CPU/VM/storage overhead, and D16/D32/D48/D64/D96 packing. Each packing case accepts its maximum replica count and rejects one more.
* Pending-pod acceptance cases exercise both scriptless and bootstrap-client provisioning. Scriptless coverage checks actual generated kubelet flags for maxPods 15/30 and the default 110/250 settings, including 100 MiB eviction and the small-node cap.
* Removed the explicit Kubernetes-version/cache table following maintainer feedback, since versions no longer select a reservation policy. Existing MaxPods tests and the shared-helper/bootstrap reservation coverage remain. The 8 GiB/110-pod unit case correctly caps the reservation at `min(2250, 8192/4) = 2048 MiB = 2 GiB`; the uncapped 64 GiB/110-pod case still expects 2250 MiB.
* Linux-target `go vet` passed for all five packages, and the changed Go files are gofmt-clean.
* Repository-wide `make verify` could not run in this WSL environment (`make: Permission denied`). Maintainer CI and live AKS validation remain required; these local tests created no real Azure resources and did not deploy to managed NAP.
* The previous revision's [E2E matrix](https://github.com/Azure/karpenter-provider-azure/actions/runs/35638489682) failed during test-cluster creation, before the suites ran. The Utilization job reports managed-cluster `QuotaExceeded`; that infrastructure blocker needs to be cleared before rerunning live E2E.

**Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note
Use maxPods-based memory reservations and a 100 MiB hard-eviction threshold consistently across all non-hardened AKS provisioning modes. Scriptless provisioning also applies these defaults to newly created nodes. Preserve node-hardening, CPU, system/storage reservations, and the VM-memory safety margin.
```

### CODE REVIEW

File: pkg/providers/instancetype/instancetypes.go [hunk 1]
```diff
@@ -182,6 +191,25 @@ func (p *DefaultProvider) List(
...
+	}
+	kubernetesVersion, err := nodeClass.GetKubernetesVersion()
+	if err != nil {
+		return false, err
+	}
+	version, err := semver.ParseTolerant(strings.TrimPrefix(kubernetesVersion, "v"))
+	if err != nil {
+		return false, fmt.Errorf("parsing Kubernetes version %q, %w", kubernetesVersion, err)
+	}
+	return version.GE(semver.Version{Major: 1, Minor: 29}), nil
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/instancetype/instancetypes.go]: Agreed. `1.29.0-azure` sorts below `1.29.0` under SemVer even though both must select the same AKS reservation policy. Fixed in 723c0dd3 by comparing the parsed major/minor, while preserving parsing/readiness validation and the provisioning-mode/hardening guards.

I reproduced four failing boundary cases before the fix. Added coverage for vendor suffixes, a leading `v`, older/newer versions, build metadata, and future major-version ordering. The three-mode acceptance tests also verify that the suffixed boundary invalidates the legacy cache entry and shares the modern entry with the stable version. All 10 unit tests (66 subtests) and 251 acceptance cases pass; one existing skip and one existing pending case remain.

**Review by theunrepentantgeek — COMMENTED** [review]
This largely looks good to me. I wonder though whether we should keep things simple and just use the new limits unconditionally.

AKS currently supports Kubernetes 1.34+ for regular use, and 1.31+ for LTS, so a threshold based on 1.29+ will always trigger.

File: pkg/providers/instancetype/instancetypes.go [hunk 2]
```diff
@@ -182,6 +191,26 @@ func (p *DefaultProvider) List(
 	return append([]*cloudprovider.InstanceType{}, result...), nil
 }
 
+func usesAKSMemoryReservations(ctx context.Context, nodeClass *v1beta1.AKSNodeClass) (bool, error) {
+	opts := options.FromContext(ctx)
+	// Scriptless explicitly configures the legacy reservations. The other modes
+	// delegate kubelet reservations to AKS; hardened nodes have a separate policy.
+	if opts.ShouldUseNodeHardening() ||
```
Comment (matthchr) [inline pkg/providers/instancetype/instancetypes.go]: Why not change the logic for scriptless too, so they're in-sync?

File: pkg/providers/instancetype/instancetype.go [hunk 3]
```diff
@@ -146,6 +146,12 @@ func newInstanceType(
 			),
 		},
 	}
+	if params.UseAKSMemoryReservations && !enableNodeHardening {
+		instanceType.Overhead.KubeReserved[corev1.ResourceMemory] = *resource.NewQuantity(
```
Comment (matthchr) [inline pkg/providers/instancetype/instancetype.go]: Agree with @theunrepentantgeek's recommendation as well: Rather than have this in a branch after the `KubeReservedResources` function, I think just roll this into `KubeReservedResources` and do it all the time. That avoids there being _3_ codepaths (1 nodeHardening, one <1.29, one >=1.29), especially as he says since 1.29+ is out of support at this point.

File: pkg/providers/instancetype/instancetypes.go (hunk 2 above)
Comment (matthchr) [reply pkg/providers/instancetype/instancetypes.go]: I think if you do the above and agree with making the change for all (not doign 1.29+), this whole method goes away since we just always do this.

Comment (GuetaHen) [reply pkg/providers/instancetype/instancetypes.go]: Agreed; addressed in bd22c959. Removed the Kubernetes-version selector and its cache flag, and applied the same non-hardened defaults to all four provisioning modes, including scriptless. Scriptless now uses the shared reservations in both its forecast and generated kubelet configuration, including the base eviction fallback. Added all-mode/maxPods cache coverage and pending-pod tests that verify the actual scriptless flags; the five affected packages pass locally, including 600 Ginkgo cases. Node hardening and the VM-memory margin remain unchanged.

File: pkg/providers/instancetype/instancetype.go (hunk 3 above)
Comment (GuetaHen) [reply pkg/providers/instancetype/instancetype.go]: Agreed with both of you; simplified this in bd22c959. The maxPods-based calculation now lives directly in KubeReservedResources, the shared non-hardened eviction default is 100Mi, and the legacy memory brackets plus post-construction overrides are gone. There are now just the standard and hardened policies. Hardening, CPU/system/storage reservations, and the 7.5% VM-memory margin are unchanged. The shared-helper regressions failed before the fix and pass afterward; all five affected Linux package suites and go vet pass. Updated the README, PR description, and release note to make the scriptless bootstrap change explicit.

File: pkg/providers/instancetype/instancetype.go [hunk 4]
```diff
@@ -373,7 +349,9 @@ func SystemReservedResources(totalMemoryMiB int64, networkPlugin string, enableN
 }
 
 func KubeReservedResources(vcpus, totalMemoryMiB int64, maxPods int32, enableNodeHardening bool) corev1.ResourceList {
-	reservedMemoryMiB := int64(1024 * reservedMemoryTaxGi.Calculate(float64(totalMemoryMiB)/1024))
+	// AKS reserves 20 MiB per pod plus 50 MiB, capped at 25% of node memory.
+	// https://learn.microsoft.com/azure/aks/node-resource-reservations#memory-reservations
+	reservedMemoryMiB := min(20*int64(maxPods)+50, totalMemoryMiB/4)
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/instancetype/instancetype.go:354]: This refers to the earlier PR contract. [Maintainers requested](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065092272) removing the Kubernetes-version gate and moving the calculation directly into KubeReservedResources, with scriptless included. Supported AKS versions use the modern policy; the pre-1.29 compatibility branch is intentionally removed. New scriptless nodes also receive the updated kube-reserved and eviction flags, so this is not merely a more optimistic forecast of unchanged bootstrap settings. The current description and tests reflect that revised scope.

File: pkg/providers/instancetype/instancetypes.go [hunk 5]
```diff
@@ -57,7 +57,7 @@ const (
 	skuRetirementHorizonMonths = 6
 )
 
-// instanceTypeParameters contains the resolved set of AKSNodeClass fields that affect
+// instanceTypeParameters contains the resolved parameters that affect
 // instance-type construction. The instance-type cache key is derived by hashing this
 // struct; adding a new field here automatically incorporates it into the key.
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/instancetype/instancetypes.go:63]: There is no longer a version-dependent reservation profile: [the maintainer requested](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065103918) removing the selector and always using the modern non-hardened policy. Consequently, changing only the version string must not create a different overhead/cache entry under this policy. maxPods remains in the hashed parameters, and the four-mode cache tests verify that changing it updates the reservation and applies the small-node cap. Reintroducing a version-policy key and readiness dependency would restore the behavior the maintainers asked to remove.

File: pkg/providers/imagefamily/bootstrap/staticvalues.go [hunk 6]
```diff
@@ -66,7 +66,7 @@ func getBaseKubeletFlags() map[string]string {
 		"--cluster-domain":                    "cluster.local",
 		"--enforce-node-allocatable":          "pods",
 		"--event-qps":                         "0",
-		"--eviction-hard":                     "memory.available<750Mi,nodefs.available<10%,nodefs.inodesFree<5%",
+		"--eviction-hard":                     "memory.available<100Mi,nodefs.available<10%,nodefs.inodesFree<5%",
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/imagefamily/bootstrap/staticvalues.go]: The PR description has been updated to the [maintainer-requested scope](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065051049): scriptless is intentionally included. Its base fallback must therefore also be 100Mi; keeping 750Mi would leave an inconsistent non-hardened bootstrap path. TestBaseKubeletFlagsHardEviction covers this default, and the existing hardened override tests plus rendered scriptless kubelet-flag tests pass. The current description and release note no longer promise the legacy scriptless fallback.

File: pkg/providers/instancetype/suite_test.go [hunk 7]
```diff
@@ -219,6 +219,86 @@ var _ = Describe("InstanceType Provider", func() {
...
+				customData := ExpectDecodedCustomData(provisionEnv)
+				ExpectKubeReservedResources(customData, "cpu=100m", "memory=650Mi", "pid=1000")
+				ExpectHardEvictionThresholds(customData, "100Mi")
+			}
+		},
+			Entry("scriptless", consts.ProvisionModeAKSScriptless),
+			Entry("bootstrap client", consts.ProvisionModeBootstrappingClient),
+		)
+
+		DescribeTable("should cache reservations by maxPods independently of the Kubernetes version", func(provisionMode string) {
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/instancetype/suite_test.go]: Ignoring the Kubernetes version in reservation selection is now the intended behavior, per [the maintainer's request](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065103918). This is an instance-type cache test, not an overall NodeClass-readiness test. It checks that maxPods changes produce the correct reservation/cap and that version-only changes reuse the entry. The obsolete version-selector tests were intentionally removed; shared-helper, bootstrap, packing, and hardening regressions cover the revised policy.

File: README.md [hunk 8]
```diff
@@ -48,6 +49,24 @@ Karpenter provider for AKS can be used in two modes:
...
 * **Self-hosted mode**: Karpenter is run as a standalone deployment in the cluster. This mode is useful for advanced users who want to customize or experiment with Karpenter's deployment, use custom Helm charts, or integrate non-standard workflows. Self-hosted mode requires users to directly manage upgrades, token rotation, and helm charts.
 
+## Node memory reservations
+
+All provisioning modes (`aksscriptless`, `bootstrappingclient`, `aksmachineapi`,
+and `aksmachineapiheaderbatch`) use the same memory reservation policy for
+non-hardened nodes: `min(20 * maxPods + 50, totalMemoryMiB / 4)` MiB of
+kube-reserved memory and a 100 MiB hard-eviction threshold. `totalMemoryMiB`
+is the nominal SKU memory in MiB. This policy applies without a Kubernetes
+version gate and follows the
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply README.md:59]: Addressed by updating the PR title, description, and release note to the [maintainer-approved scope](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065092272): one modern non-hardened policy for all provisioning modes, including generated kubelet settings on new scriptless nodes, with no Kubernetes-version gate. The README now agrees with that description. Node hardening and the separate VM-memory margin remain unchanged.

File: pkg/providers/instancetype/instancetype.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthchr) [reply pkg/providers/instancetype/instancetype.go:354]: This behavior matches Machine API and 1.29 is not supported anymore.

File: pkg/providers/instancetype/instancetypes.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (matthchr) [reply pkg/providers/instancetype/instancetypes.go:63]: No longer considering 1.29+ so this should matter anymore?

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (matthchr) [reply README.md:59]: I think just correct the PR description

File: pkg/providers/imagefamily/bootstrap/staticvalues.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthchr) [reply pkg/providers/imagefamily/bootstrap/staticvalues.go:69]: As mentioned in other comment, IMO fix here is just change PR description, you are doing what we asked =)

File: pkg/providers/instancetype/instancetype.go [hunk 9]
```diff
@@ -40,34 +40,10 @@ import (
 
 const (
 	MemoryAvailable        = "memory.available"
-	DefaultMemoryAvailable = "750Mi"
+	DefaultMemoryAvailable = "100Mi"
```
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/instancetype/instancetype.go:43]: This is the intentional scope change requested by the maintainers: [include scriptless](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065051049) and [use the shared policy unconditionally](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065092272). The forecast and generated kubelet settings both change for new scriptless nodes; existing nodes are not reconfigured, and hardened thresholds remain separate. The current PR description and release note document this. Restoring the 750Mi/version split would undo that requested change.

File: pkg/providers/imagefamily/bootstrap/staticvalues.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (GuetaHen) [reply pkg/providers/imagefamily/bootstrap/staticvalues.go:69]: The PR description has been updated to the [maintainer-requested scope](https://github.com/Azure/karpenter-provider-azure/pull/1919#discussion_r4065051049): scriptless is intentionally included. Its base fallback must therefore also be 100Mi; keeping 750Mi would leave an inconsistent non-hardened bootstrap path. TestBaseKubeletFlagsHardEviction covers this default, and the existing hardened override tests plus rendered scriptless kubelet-flag tests pass. The current description and release note no longer promise the legacy scriptless fallback.

File: pkg/providers/instancetype/suite_test.go (hunk 7 above)
Comment (matthchr) [inline pkg/providers/instancetype/suite_test.go]: Since k8s version doesn't matter anymore maybe you can just remove this test rather than have it explicitly?

**Review by matthchr — DISMISSED** [review]

File: pkg/providers/instancetype/suite_test.go (hunk 7 above)
Comment (GuetaHen) [reply pkg/providers/instancetype/suite_test.go]: Removed in e295fc51. The explicit version/cache table is gone. Existing MaxPods tests and the reservation-cap, bootstrap, packing, and hardening coverage remain. Re-ran the full instance-type suite: 249 Ginkgo cases plus its Go unit tests passed. Production code is unchanged.

**Review by matthchr — APPROVED** [review]
Approved, but waiting to merge until I can get an E2E run on this as well.

**Comment by matthchr** [comment]
E2E looked good. Here's a fresh run on scriptless: https://github.com/Azure/karpenter-provider-azure/actions/runs/35778645882

edit: some failures due to metrics server placement. I am re-running to ensure they pass, then will merge.
